### PR TITLE
Rework reading pod's information for resource monitor plugin

### PR DESCRIPTION
**Pull request by svor**
Signed-off-by: svor <vsvydenk@redhat.com>

### What does this PR do?
Rework the process of reading pod's information for resource monitor plugin

### Screenshot/screencast of this PR
Tested on local minikube and on OCP 4.6.24

screenshot-nimbus-capture-2021 04 22-13_19_27

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/18812

### How to test this PR?
This PR depends on https://github.com/eclipse-che/che-operator/pull/793 and https://github.com/eclipse/che/pull/19651

To test this PR need to deploy che with custom `che-operator` image which should include changes from https://github.com/eclipse-che/che-operator/pull/793 and custom `cheimage` which should include changes from https://github.com/eclipse/che/pull/19651. 

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [x] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [x] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [x] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [x] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [x] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [x] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### Happy Path Channel

HAPPY_PATH_CHANNEL=stable

### CODE REVIEW

File: plugins/resource-monitor-plugin/src/resource-monitor-plugin.ts [hunk 1]
```diff
@@ -193,6 +185,6 @@ export class ResMon {
 }
 
 export async function getNamespace(): Promise<string> {
-  const devfile = await che.devfile.get();
-  return devfile.metadata?.attributes ? devfile.metadata.attributes.infrastructureNamespace : '';
+  const workspace = await che.workspace.getCurrentWorkspace();
+  return workspace.attributes?.['infrastructureNamespace'] || workspace.namespace || '';
```
Comment (benoitf) [inline plugins/resource-monitor-plugin/src/resource-monitor-plugin.ts:189]: hi, `workspace.attributes` is only a che-server implementation (so we should keep devfileService to grab the attributes information)

File: extensions/eclipse-che-theia-plugin-ext/src/plugin/che-api.ts [hunk 2]
```diff
@@ -433,6 +433,10 @@ export function createAPIFactory(rpc: RPCProtocol): CheApiFactory {
       sendRawQuery(requestURL: string, opts: any): Promise<K8SRawResponse> {
         return cheK8SImpl.sendRawQuery(requestURL, opts);
       },
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      getWorkspacePod(): Promise<any> {
```
Comment (benoitf) [inline extensions/eclipse-che-theia-plugin-ext/src/plugin/che-api.ts:437]: here it might be better to send a real type instead of any

**Review by benoitf — COMMENTED** [review]
hi, why the plug-in can't get the pod using the kubernetes API instead of adding a getPod method to the che API ?

**Comment by svor** [comment]
hello @benoitf, thank you for the review. 
As I remember we decided to not add k8s API library as a dependency into the plugin to avoid some extra memory requirements. Why do you find it as a problemm I mean to add a method into che API. The implementation of this method is using k8s API and all plugins can reuse this method easely

**Comment by benoitf** [comment]
hi @svor what is worrying me is that we return undefined any type so it's hard to know from consumer what you're handling and how to know what is inside this struct.

if a plug-in uses the kubernetes API then they can grab any information they want from the cluster, not only pods

**Comment by svor** [comment]
> hi @svor what is worrying me is that we return undefined any type so it's hard to know from consumer what you're handling and how to know what is inside this struct.

It is not a problem, we can return an object, I'll rework it

File: plugins/resource-monitor-plugin/src/resource-monitor-plugin.ts (hunk 1 above)
Comment (svor) [reply plugins/resource-monitor-plugin/src/resource-monitor-plugin.ts:189]: @benoitf I've tried to get namespace from devfileService and `devfile.metadata?.attributes` returns
```json
{"metadata-name-field":"name"}
```
(no `infrastructureNamespace` here)
but if use workspaceService, `workspace.attributes?.['infrastructureNamespace']` returns real namespace: `admin-che`

Comment (benoitf) [reply plugins/resource-monitor-plugin/src/resource-monitor-plugin.ts:189]: maybe there is a bug on how we transmit properties into devfile service but we can't rely on workspace.attributes field when using devWorkspace so we shouldn't introduce this dependency that was removed

Comment (svor) [reply plugins/resource-monitor-plugin/src/resource-monitor-plugin.ts]: Fixed. Now `infrastructureNamespace` attribute is reading from devfle service.

**Review by benoitf — COMMENTED** [review]
Hello,

https://user-images.githubusercontent.com/436777/117274003-f2489780-ae5c-11eb-9438-88f4974b7eba.mp4


when testing, when workspace is starting, I've for some time a "No resource available" with a 404 error
So I don't know if it's because plug-in do not wait some events to start ?

**Comment by svor** [comment]
@benoitf The plugin tries to read metrics of workspace pod from Metrics server if it is alive. There are no some special events to wait, probably, when the workspace just started, Metrics server doesn't have information about workspace pod (it is not ready yet), that's why it returns 404. But when it will be available, we can see it on the status bar. I don't find it as a problem, do you think it is not user friendly?

**Comment by benoitf** [comment]
@svor yes I don't think it's user friendly so maybe we wait to display information instead of saying it's not there and we can see 404 errors (can be a follow up)

**Review by benoitf — APPROVED** [review]
user experience might be improved when starting a workspace but can be a follow-up

**Comment by artaleks9** [comment]
[crw-ci-test --rebuild]

**Comment by svor** [comment]
@benoitf what do you think about this version, is it better:
https://user-images.githubusercontent.com/1271546/117302914-d0630b00-ae84-11eb-8083-0c1730bb1567.mp4

**Comment by benoitf** [comment]
@svor yes much better, also if we do know metrics are coming, we could display `Initializing` or `Waiting metrics...` in the output (and/or display spinner icon)

**Pull request merged**